students/HarveyHerela/lesson04/trigram.py

```diff
@@ -0,0 +1,211 @@
+import string
+import random
+
+# Trigram is dictionary where the keys are 2-word tuples and the
+# values are a list of words
+trigrams = {}
+
+# For creating the trigrams
+prev_words = []
+
+# Name of the book to parse
+book_filename = "jack_london_white_fang.txt"
+
+def add_word_to_trigram(new_word):
+    """Helper function for creating the trigrams"""
+    # Trigrams require 2 previous words
+    # If we don't have those yet, then set them
+    if len(prev_words) < 2:
+       prev_words.append(new_word)
+       return
+
+    # If it exists, add the word to the list
+    # If it doesn't exist, create it
+    word_tuple = (prev_words[0], prev_words[1])
+    if word_tuple in trigrams:
+        trigrams[word_tuple].append(new_word)
+    else:
+        trigrams[word_tuple] = [new_word]
+
+    # Increment the prev words
+    prev_words.pop(0)
+    prev_words.append(new_word)
+
+def break_it_down(line):
+    """Separates 'line' by spaces, similar to the built-in 'split' function.
+    However, it also treats punctuation as separate words.
+    """
+    # Make a copy for us to play with
+    words = str(line)
+    # Strategy: Leverage the built-in split, by adding spaces
+    # around punctuation
+    # Supported punctuation: ; : , ? ! . "
+    for p in (':', ';', ',', '?', '!', '.', '"'):
+        words = words.replace(p, ' {p} '.format(p=p))
+    
+    # Special case: ellipsis
+    while '.   .' in words:
+        words = words.replace('.   .', '..')
+        
+    # Remove this punctuation: ( )
+    for p in ('(', ')'):
+        words = words.replace(p, ' ')
+    
+    return words.split()
+
+def read_in_book():
+    """Reads in a book and creates the trigrams"""
+    # Keep track of paragraphs
+    new_lines = False
+    with open(book_filename, 'r') as book:
+        # Ignore the top of the file, it's all boilerplate
+        # 'project gutenberg' stuff anyway
+        line_it = iter(book)
+        line = next(line_it)
+        while line[:3] != "***":
+            line = next(line_it)
+
+        line = next(line_it)
+
+        # Read the rest, creating trigrams, until
+        # another series of *** are found
+        while line[:3] != "***":
+            words = break_it_down(line)
+            if len(words) == 0:
+                new_lines = True
+            else:
+                if new_lines:
+                    add_word_to_trigram('\n\n')
+                new_lines = False
+                
+            for w in words:
+                add_word_to_trigram(w)
+
+            line = next(line_it)
```

Comment:
I think this could be simplified to:
```
with open(book_filename, 'r') as book:
    for line in book:
        if line[:3] == "***":
            continue
        else:
            process_line(line)
```